Mathlib/LinearAlgebra/TensorProduct/Submodule.lean

```diff
@@ -0,0 +1,244 @@
+
+theorem mulMap_comp_rTensor (M' : Submodule R S) (hM : M' ≤ M) :
+    mulMap M N ∘ₗ (inclusion hM).rTensor N = mulMap M' N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_lTensor (N' : Submodule R S) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ (inclusion hN).lTensor M = mulMap M N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_map_inclusion (M' N' : Submodule R S) (hM : M' ≤ M) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ TensorProduct.map (inclusion hM) (inclusion hN) = mulMap M' N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_eq_mul'_comp_mapIncl : mulMap M N = .mul' R S ∘ₗ TensorProduct.mapIncl M N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_range : LinearMap.range (mulMap M N) = M * N := by
+  refine le_antisymm ?_ (mul_le.2 fun m hm n hn ↦ ⟨⟨m, hm⟩ ⊗ₜ[R] ⟨n, hn⟩, rfl⟩)
+  rintro _ ⟨x, rfl⟩
+  induction x using TensorProduct.induction_on with
+  | zero => rw [_root_.map_zero]; exact zero_mem _
+  | tmul a b => exact mul_mem_mul a.2 b.2
+  | add a b ha hb => rw [_root_.map_add]; exact add_mem ha hb
+
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] M * N` induced by multiplication in `S`,
+which is surjective (`Submodule.mulMap'_surjective`). -/
+def mulMap' : M ⊗[R] N →ₗ[R] ↥(M * N) :=
+  (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
+
+@[simp]
+theorem val_mulMap'_tmul (m : M) (n : N) : (mulMap' M N (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
+
+theorem mulMap'_surjective : Function.Surjective (mulMap' M N) := by
+  simp_rw [mulMap', LinearMap.coe_comp, LinearEquiv.coe_coe, EquivLike.comp_surjective,
+    LinearMap.surjective_rangeRestrict]
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural map
+`i(R) ⊗[R] N →ₗ[R] N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This is promoted to an isomorphism as `Submodule.lTensorOne`. Use that instead. -/
+def lTensorOne' : (⊥ : Subalgebra R S) ⊗[R] N →ₗ[R] N :=
+  show (1 : Submodule R S) ⊗[R] N →ₗ[R] N from
+    (LinearEquiv.ofEq _ _ (by rw [mulMap_range, one_mul])).toLinearMap ∘ₗ (mulMap _ N).rangeRestrict
+
+@[simp]
+theorem lTensorOne'_tmul (y : R) (n : N) :
+    N.lTensorOne' (algebraMap R _ y ⊗ₜ[R] n) = y • n := Subtype.val_injective <| by
+  simp_rw [lTensorOne', LinearMap.coe_comp, LinearEquiv.coe_coe, Function.comp_apply,
+    LinearEquiv.coe_ofEq_apply, LinearMap.codRestrict_apply, SetLike.val_smul, Algebra.smul_def]
+  exact mulMap_tmul 1 N _ _
+
+@[simp]
+theorem lTensorOne'_one_tmul (n : N) : N.lTensorOne' (1 ⊗ₜ[R] n) = n := by
+  simpa using lTensorOne'_tmul N 1 n
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural isomorphism between
+`i(R) ⊗[R] N` and `N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This generalizes `TensorProduct.lid` as `i(R)` is not necessarily isomorphic to `R`. -/
+def lTensorOne : (⊥ : Subalgebra R S) ⊗[R] N ≃ₗ[R] N := by
```

Comment:
Can you define this in term mode? (Proving the required `Prop` in tactic mode is of course fine.)

Reply:
Done. (In fact it doesn't matter, since I've checked that `by refine LinearEquiv.ofLinear ...` actually compiles into `LinearEquiv.ofLinear ...`.)

Reply:
Ah OK, I wasn't sure about that.